driver-core/src/main/com/mongodb/internal/connection/AsynchronousClusterEventListener.java

```diff
@@ -0,0 +1,172 @@
+        publishingThread.setDaemon(true);
+        publishingThread.start();
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    Thread getPublishingThread() {
+        return publishingThread;
+    }
+
+    @Override
+    public void clusterOpening(final ClusterOpeningEvent event) {
+        addClusterEventInvocation(clusterListener -> clusterListener.clusterOpening(event), false);
+    }
+
+    @Override
+    public void clusterClosed(final ClusterClosedEvent event) {
+        addClusterEventInvocation(clusterListener -> clusterListener.clusterClosed(event), true);
+    }
+
+    @Override
+    public void clusterDescriptionChanged(final ClusterDescriptionChangedEvent event) {
+        addClusterEventInvocation(clusterListener -> clusterListener.clusterDescriptionChanged(event), false);
+    }
+
+    @Override
+    public void serverOpening(final ServerOpeningEvent event) {
+        addServerEventInvocation(serverListener -> serverListener.serverOpening(event));
+    }
+
+    @Override
+    public void serverClosed(final ServerClosedEvent event) {
+        addServerEventInvocation(serverListener -> serverListener.serverClosed(event));
+    }
+
+    @Override
+    public void serverDescriptionChanged(final ServerDescriptionChangedEvent event) {
+        addServerEventInvocation(serverListener -> serverListener.serverDescriptionChanged(event));
+    }
+
+    @Override
+    public void serverHearbeatStarted(final ServerHeartbeatStartedEvent event) {
+        addServerMonitorEventInvocation(serverMonitorListener -> serverMonitorListener.serverHearbeatStarted(event));
+    }
+
+    @Override
+    public void serverHeartbeatSucceeded(final ServerHeartbeatSucceededEvent event) {
+        addServerMonitorEventInvocation(serverMonitorListener -> serverMonitorListener.serverHeartbeatSucceeded(event));
+    }
+
+    @Override
+    public void serverHeartbeatFailed(final ServerHeartbeatFailedEvent event) {
+        addServerMonitorEventInvocation(serverMonitorListener -> serverMonitorListener.serverHeartbeatFailed(event));
+    }
+
+    private void addClusterEventInvocation(final VoidFunction<ClusterListener> eventPublisher, final boolean isLastEvent) {
+        addEvent(() -> {
+            eventPublisher.apply(clusterListener);
+            return isLastEvent;
+        });
+    }
+
+    private void addServerEventInvocation(final VoidFunction<ServerListener> eventPublisher) {
+        addEvent(() -> {
+            eventPublisher.apply(serverListener);
+            return false;
+        });
+    }
+
+    private void addServerMonitorEventInvocation(final VoidFunction<ServerMonitorListener> eventPublisher) {
+        addEvent(() -> {
+            eventPublisher.apply(serverMonitorListener);
+            return false;
+        });
+    }
+
+    private void addEvent(final Supplier<Boolean> supplier) {
+        // protect against rogue publishers
+        if (!publishingThread.isAlive()) {
+            return;
+        }
```

Comment:
> server monitor events can be published after cluster is closed because DefaultServerMonitor#close only interrupts the monitorThread, but does not join it

Thanks! That means that the calling code is racy and we are fine with ignoring the events that arrive after the thread is terminated.

The check `if (!publishingThread.isAlive())` is still there, what purpose does it serve given the observations from my previous comment?

Reply:
I guess it's a little bit defensive programming, which I generally avoid.  The idea is that if for whatever reason the monitor thread doesn't stop producing events even after the cluster closes, we won't keep adding to the queue with nothing to empty it.

Reply:
Given that, do you want me to remove it?

Reply:
> if for whatever reason the monitor thread doesn't stop producing events even after the cluster closes

I see. I would not think about doing this, but since defending against that is trivial and the code has already been written, let's leave it as is.